derp/derp_server.go

```diff
@@ -1942,6 +2101,111 @@ func (s *Server) ServeDebugTraffic(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// takeFlowsSnapshot returns a map of copies of all the flowStats for
+// this server. The output should be suitable for comparing the stats
+// for a specific flow at multiple points in time so that we can
+// calculate the bandwidth used.
+func (s *Server) takeFlowsSnapshot() map[*sclient]map[key.NodePublic]map[key.NodePublic]flowStats {
+	snap := make(map[*sclient]map[key.NodePublic]map[key.NodePublic]flowStats)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for c := range s.flows {
+		snap[c] = make(map[key.NodePublic]map[key.NodePublic]flowStats)
+		for dstKey := range s.flows[c] {
+			snap[c][dstKey] = make(map[key.NodePublic]flowStats)
+			for srcOrFwd := range s.flows[c][dstKey] {
+				snap[c][dstKey][srcOrFwd] = *s.flows[c][dstKey][srcOrFwd]
+			}
+		}
+	}
+	return snap
+}
+
+// ServeFlows prints a list of recently active data flows through
+// this server, including the bandwidth used by each over the last
+// second.
+func (s *Server) ServeFlows(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "current time: %s\n", time.Now().Format("2006-01-02 15:04:05"))
+
+	lastTime := time.Now()
+	lastSnap := s.takeFlowsSnapshot()
+	time.Sleep(1 * time.Second)
+	currTime := time.Now()
+	currSnap := s.takeFlowsSnapshot()
+	dur := currTime.Sub(lastTime)
```

Comment:
A comment explaining why we're taking two snapshots and sleeping for 1 second would help here; I think it's because we're using the `lastSnap` to calculate bandwidth over the 1s interval, but I'm not sure I'm reading this code right? Does the fact that it's 1s matter, other than making the "per-second" column calculations easier?

Reply:
Thanks, comment added! 1 second is an arbitrary "enough" to get some kind of reasonable bandwidth estimate while not delaying too long in response.